src/event-engine/states/handshake_failed.ts

```diff
@@ -0,0 +1,42 @@
+import { State } from '../core/state';
+import { Effects } from '../effects';
+import { Events, reconnect, restore, subscriptionChange, unsubscribeAll } from '../events';
+import { PubNubError } from '../../core/components/endpoint';
+import { HandshakingState } from './handshaking';
+import { UnsubscribedState } from './unsubscribed';
+import { Cursor } from '../../models/Cursor';
+
+export type HandshakeFailedStateContext = {
+  channels: string[];
+  groups: string[];
+  cursor?: Cursor;
+
+  reason: PubNubError;
+};
+
+export const HandshakeFailedState = new State<HandshakeFailedStateContext, Events, Effects>('HANDSHAKE_FAILED');
+
+HandshakeFailedState.on(subscriptionChange.type, (_, event) => {
+  return HandshakingState.with({ channels: event.payload.channels, groups: event.payload.groups });
+});
```

Comment:
It looks like in this case we will throw away `cursor` which user potentially could pass in the other state, and we were keeping it for transition to the `Receiving` state.

```suggestion
HandshakeFailedState.on(subscriptionChange.type, (context, event) => {
  return HandshakingState.with({ 
    channels: event.payload.channels, 
    groups: event.payload.groups,
    cursor: context.cursor,
  });
});
```

Reply:
✔️